[RFC] Fix os_file_exists to not approve directories

danobi opened the pull request:
As described in issue #4939

ZyX-I commented:
This will break `vim_rename` and who knows what else. Proposed fix was to fix the name, not behaviour.

danobi commented:
@ZyX-I Sorry, I misunderstood. 

How does this look now? I renamed `os_file_exists` to `os_path_exists` and updated all the call sites. I also created a function called `os_is_reg_file` that checks if a path points to a regular file, since it looked like that function didn't exist already.

ZyX-I commented:
@danobi It is not a library for third-party uses, `os_is_reg_file` is not needed until it is used. If you want to write tests for it, I have nothing against, but so far this looks like a waste of time.

I would also suggest adding tests regarding what `os_path_exists` returns for directory, symbolic link and broken symbolik link (latter is not created AFAIK, but is easy to: `lfs.link('xxx_unexistent_file_xxx', 'unit-test-directory/test_broken_link.file', true)`).

danobi commented:
@ZyX-I Does this look reasonable?

ZyX-I commented:
This PR LGTM (as its purpose is renaming `os_file_exists` if I am not mistaking), but apparently `vim_rename` has a bug: basically when it uses `os_file_exists` (`os_path_exists` now) it wants to check whether it may use some location to copy file to (i.e. it checks whether location is *occupied*). `os_path_exists` will return `false` for broken symbolic link which may either delete that broken symlink (most likely) or fail without trying any other possible temporary locations (I actually do not know condition under which behaviour will be exactly this). Though Python’s `os.path.exists` does the same thing for broken symlinks (intentional, such behaviour is described in its documentation). Not sure whether this needs to be fixed.

ZyX-I commented:
Do not forget to silence linter though.

justinmk commented:
QB is failing because of me, ignore it. I am looking at it now.

justinmk commented on src/nvim/fileio.c:
```diff
@@ -4514,9 +4515,14 @@ int vim_rename(char_u *from, char_u *to)
+      // between the tail pointer and the start of the array from the
+      // total buffer length
+      char * tail = (char *)path_tail(tempname);
+      snprintf(tail, (MAXPATHL + 1) - (tail - (char *)tempname - 1), "%d", n);
```
the comment is too long, and the mention of "n" is confusing given the presence of the `n` parameter which is unrelated to the 'n' mentioned in the comment. Better to leave the comment out, it mostly restates what the code is doing.

danobi commented:
Done. Should I be worried about the travis builds? It looks like the failing builds are timing out.

justinmk commented:
the hanging macOS builds are unrelated, dealing with them elsewhere. Please squash and will merge!

danobi committed:
os/fs: Rename os_file_exists to os_path_exists

Because the old name did not indicate that the function
would return true on directories as well.

Changes:
  - function name changed

  - all call sites updated

  - new tests added for broken symlinks and directories

  - surrounding linter warnings resolved

danobi commented:
Squashed

justinmk commented:
Thanks! Anything in the [current milestone](https://github.com/neovim/neovim/milestone/10) would help a lot. The [static analysis](https://neovim.io/doc/reports/clang/) issues also need to be fixed.
